torch/_functorch/aot_autograd.py

```diff
@@ -2772,6 +3038,277 @@ def traced_forward(*primals_fwd_seed_fwd_base_offset):
+    # These activations are not visible to the user, and so there's no need for us to wrap them back into subclasses.
+    #
+    # On top of that, when we first computed subclass metadata (in `run_functionalized_fw_and_collect_metadata`),
+    # we computed subclass metadata on every forward output, but this did **not** include activations
+    # created by the partitioner.
+    # as a result, `unwrapped_args` here will correspond to (*unwrapped_user_fw_outs, *activations),
+    # but `subclass_metas` will only correspond to subclass metatadata on `user_fw_outs`.
+    # We then need to make sure that we return (*wrapped_user_fw_outs, *activations).
+    if num_fw_outs_saved_for_bw is not None:
+        assert len(unwrapped_args) == num_args_tallied + num_fw_outs_saved_for_bw
+        activations = unwrapped_args[num_args_tallied:]
+        if isinstance(wrapped_args, tuple) and isinstance(activations, tuple):
+            return wrapped_args + activations
+        return tuple(list(wrapped_args) + list(activations))
+    else:
+        assert len(unwrapped_args) == num_args_tallied
+        return tuple(wrapped_args)
+
+# Given a bunch of "dense" tensor arguments, this function (potentially) wraps them into tensor subclasses.
+# This function carefully handles the inference vs. joint cases:
+# - when trace_joint is True, args is (primals, tangents)
+# - when trace_joint is False, args is [*primals]
+def wrap_tensor_subclasses(unwrapped_args, *, trace_joint: bool, meta: ViewAndMutationMeta) -> List[Any]:
+    # Since this function is re-used for both inference and joint graphs,
+    if trace_joint:
+        assert isinstance(unwrapped_args, tuple) and len(unwrapped_args) == 2
+        assert isinstance(unwrapped_args[0], (tuple, list)) and isinstance(unwrapped_args[1], (tuple, list))
+        primals, tangents = unwrapped_args[0], unwrapped_args[1]
+        wrapped_primals = wrap_tensor_subclasses_simple(primals, subclass_metas=meta.subclass_inp_meta)
+        wrapped_tangents = wrap_tensor_subclasses_simple(tangents, subclass_metas=meta.subclass_tangent_meta)
+        return (wrapped_primals, wrapped_tangents)
+    else:
+        wrapped_args = wrap_tensor_subclasses_simple(unwrapped_args, subclass_metas=meta.subclass_inp_meta)
+        return wrapped_args
+
+# This wrapper handles the AOTDispatch runtime logic for tensor subclasses.
+# At runtime, we have a compiled function that knows how to operate on the domain of DenseTensor -> DenseTensor,
+# But the user might have passed us some tensor subclass inputs (or expect some subclass tensor outputs).
+# This function handles the wrapping and unwrapping of tensor subclasses at runtime.
+def aot_dispatch_subclass_wrapper(
+    runtime_fn: Callable,
+    *,
+    subclass_metas: List[Union[int, SubclassCreationMeta]],
+    num_fw_outs_saved_for_bw: Optional[int],
+) -> Callable:
+    def inner_fn(*args):
+        unwrapped_args = unwrap_tensor_subclasses(args, trace_joint=False)
```

Comment:
kwarg is slightly misleading here, you might in fact be wrapping an autograd custom function here, the point is that you don't have the tuple structure in args

Reply:
i'll rename to is_joint_structure, hopefully less confusing but open to suggestions